src/miner.cpp

```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+
+    // This serializes with output value, a fixed-length 8 byte field, of zero and height, a serialized CScript
+    // signed integer taking up 4 bytes for heights 32768-8388607 (around the year 2167) after which it will use 5.
+    nCoinbaseSize = coinbaseTx(scriptPubKeyIn, 400000, 0).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION);
+    // BU005 END
+
+    // BU Miners take the block we give them, wipe away our coinbase and add their own.
+    // So if their reserve choice is bigger then our coinbase then use that.
+    return nHeaderSize + std::max(nCoinbaseSize, coinbaseReserve.value);
+}
+
+CMutableTransaction BlockAssembler::coinbaseTx(const CScript &scriptPubKeyIn, int nHeight, CAmount nValue)
+{
+    CMutableTransaction tx;
+
+    tx.vin.resize(1);
+    tx.vin[0].prevout.SetNull();
+    tx.vout.resize(1);
+    tx.vout[0].scriptPubKey = scriptPubKeyIn;
+    tx.vout[0].nValue = nValue;
+    tx.vin[0].scriptSig = CScript() << nHeight << OP_0;
+
+    // BU005 add block size settings to the coinbase
+    std::string cbmsg = FormatCoinbaseMessage(BUComments, minerComment);
+    const char *cbcstr = cbmsg.c_str();
+    vector<unsigned char> vec(cbcstr, cbcstr + cbmsg.size());
+    COINBASE_FLAGS = CScript() << vec;
+    // Chop off any extra data in the COINBASE_FLAGS so the sig does not exceed the max.
+    // we can do this because the coinbase is not a "real" script...
+    if (tx.vin[0].scriptSig.size() + COINBASE_FLAGS.size() > MAX_COINBASE_SCRIPTSIG_SIZE)
     {
-      tmpl = CreateNewBlock(chainparams, scriptPubKeyIn, true);
+        COINBASE_FLAGS.resize(MAX_COINBASE_SCRIPTSIG_SIZE - tx.vin[0].scriptSig.size());
     }
+    tx.vin[0].scriptSig = tx.vin[0].scriptSig + COINBASE_FLAGS;
 
-  return tmpl;
+    return tx;
 }
 
-CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const CScript& scriptPubKeyIn,bool blockstreamCoreCompatible)
+CBlockTemplate *BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn)
 {
-    // Create new block
-    std::unique_ptr<CBlockTemplate> pblocktemplate(new CBlockTemplate());
-    if(!pblocktemplate.get())
-        return NULL;
-    CBlock *pblock = &pblocktemplate->block; // pointer for convenience
+    CBlockTemplate *tmpl = NULL;
+    if (maxGeneratedBlock > BLOCKSTREAM_CORE_MAX_BLOCK_SIZE)
+        tmpl = CreateNewBlock(scriptPubKeyIn, false);
 
-    // Create coinbase tx
-    CMutableTransaction txNew;
-    txNew.vin.resize(1);
-    txNew.vin[0].prevout.SetNull();
-    txNew.vout.resize(1);
-    txNew.vout[0].scriptPubKey = scriptPubKeyIn;
+    // If the block is too small we need to drop back to the 1MB ruleset
+    if ((!tmpl) || (tmpl->block.nBlockSize <= BLOCKSTREAM_CORE_MAX_BLOCK_SIZE))
+    {
+        tmpl = CreateNewBlock(scriptPubKeyIn, true);
+    }
+
+    return tmpl;
+}
+
+CBlockTemplate *BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn, bool blockstreamCoreCompatible)
+{
+    resetBlock(scriptPubKeyIn);
+
+    pblocktemplate.reset(new CBlockTemplate());
```

Comment:
You have a unique pointer here with pblocktemplate. One which you fill in the start of this method and release at the end.
The problem is that you throw exceptions which avoids releasing this variable.

Why is this not a variable local to this method instead of a class member?